Add Visual Editor Docs (#248)
directus/docs: content/guides/2.content/8.visual-editor/1.frontend-package.md

Reviewer: This part is missing: https://github.com/directus/visual-editing/blob/init/readme.md#usage-for-non-js-environments

Reviewer 2: @formfcw is it fair to say this is more like usage for non-npm/yarn/etc-based environments?

Reviewer: … or for use without `JS package managers`?

Reviewer: Also customization options are missing: https://github.com/directus/visual-editing/blob/init/readme.md#customization

Reviewer 2: Can you please write me up a barebones set of docs for this? I'd need to know:

- What the first bullet is about (IDs, etc)
- An example of adding custom classes

Author: Yeah I'll add this in either through a separate PR or just straight in here. Prob fine to just be straight in as it sits separately.

Reviewer: @hola-soy-milk You might also want to take a look at this document and see what might be useful: https://github.com/directus/visual-editing/blob/init/readme.md

The `apply()` method also returns the `remove`, `enable` and `disable` methods. These can then be used on the selected elements in question. Make sure to await them.

```js
const { disable, enable, remove } = await apply({ directusUrl });
```

The optional `onSaved` callback function of the `apply()` method has a number of parameters which may be useful.

| Parameter | Type | Description |
| --------- | ---- | ----------- |
| `collection` | `string`              | Name of the relevant collection. |
| `item` | `string`, `number` | Primary key of the item |
| `payload` | `Record<string, any>` | The changed values. |

## Content Security Policy

If you have CSP configured, be sure to add `'frame-ancestors': ["'self'", process.env.DIRECTUS_URL]`, to your policy — in other words, make your site available for use inside an iframe in Directus.

If you're unsure where your CSP is defined, check your web server configuration files, your site's build configuration, or your hosting platform's security settings.

## Usage in Non-JS Environments

> As this library is built as a Node package, environments that can't take advantage of NPM - like those built with PHP - will need to take a slightly different approach to including the functionality in their websites.

Reviewer: ```suggestion
As this library is built as a Node package, environments that can't take advantage of NPM will need to take a slightly different approach to including the functionality in their websites.
```

Actually, you CAN use NPM to compile your JS or CSS files for a PHP site... so let's keep it simple and just not mention PHP.

> The compiled code for the package can be found [in the repository](https://github.com/directus/visual-editing/blob/c634a3fe4f9dde298d3ae13d3e7680b59c61ef99/dist/visual-editing.js). 

Reviewer: Thoughts on making this a call-out?

---

> title: Frontend Package

Reviewer: ```suggestion
title: Website Library
```

```suggestion
title: Website Package
```

Reviewer 2: I again thing frontend is better here

> description: An NPM package used to connect your website's frontend to your Directus instance.

Reviewer: ```suggestion
description: A library used to connect your website to your Directus instance and let them communicate with each other.
```

Maybe we should also somehow mention that this is necessary to enable Visual Editing?

```suggestion
description: A library that allows your website to communicate with your Directus instance and enables Visual Editing.
```

Reviewer 2: ```suggestion
description: A library that allows your website to communicate with your Directus project and enables Visual Editing.
```

---

> Developers must make use of the Visual Editor [NPM package](https://link-to-npm.com) to create links between your website's HTML elements and the Directus Visual Editor Module. This is done through data attributes and helper functions built into the package and imported into your website.

Reviewer: - make use of "Visual Editing"
- call it NPM package?
- The link to the npm package will be https://www.npmjs.com/package/@directus/visual-editing
- it does not create links … it makes content editable

Reviewer 2: - Sounds good! 
- What do you mean?
- Thank you!
- Doesn't it make the content editable by linking the HTML elements to the module?

Author: > What do you mean?

I think he means just call it "Visual Editing Package", no need for NPM? @formfcw

> it does not create links … it makes content editable

I meant linkages, not actual anchor links, but I agree the term is not good.

> Doesn't it make the content editable by linking the HTML elements to the module?

With my above clarification I think your thinking is correct @hola-soy-milk 👍

Reviewer: To sum it up:

```suggestion
Developers must make use of the Visual Editing [Frontend package](https://www.npmjs.com/package/@directus/visual-editing) to ensure communication between your website's HTML elements and the Directus Visual Editor Module. This is done through data attributes and helper functions built into the package and imported into your website.
```

Makes sense?

```bash

> npm install @directus/visual-editor

Reviewer: The library is called @directus/visual-editing (not ~~@directus/visual-editor~~)

```suggestion
npm install @directus/visual-editing
```

```

## API

> The Visual Editor Package consists of four imported methods, only one of which is required to make the basic functionality work.

Reviewer: If a method is added, we would need to update this number, so I suggest avoiding this.

```suggestion
The Visual Editing Package consists of a few imported methods, only one of which is required to make the basic functionality work.
```

Author: Yeah, maybe say "not all of which", or just end the sentence at "methods.""

| Method             | Description                                                                          | Options |
| ------------------ | -------------------------------------------------------------------------------------| ------- |
| `apply()`          | Connects each `data-directus` attribute to your Directus instance. This is required. | `directusUrl`, `elements`, `customClass`, `onSaved` |
| `setAttr()`        | Helper function for dynamically generating `data-directus` attributes.               | `collection`, `item`, `fields`, `mode`              |

> | `remove()`         | Removes all elements.                                                                | |

Reviewer: @hola-soy-milk I think we should mention this – from my orignial document:

> Before leaving the page on client-side navigation, be sure to call the remove method

Reviewer 2: Yup, I'll add this as a callout

| `disable()`        | Temporarily disable all elements.                                                    | |

## Configuring Attributes

The association between individual website elements and Directus collections and items is made through `data-directus` attributes rendered within the HTML of your website.

The preferred method for generating your `data-directus` attributes is by using the included `setAttr` helper method in order to render these attributes dynamically.

For example when using `setAttr` in Vue:

```vue
<template>
    <h1 :data-directus="setAttr({ collection: 'posts', item: post.id, fields: 'title',
    mode: 'popover' })">
        {{ post.title }}
    </h1>
</template>
```

> The generated attribute will be:

Reviewer: What we don’t learn here is that fields can also be arrays and that they render like

```html
<h1 data-directus="collection:posts;item:12;fields:title,subtitle,slug">
  I Love Visual Editing
</h1>
```

Reviewer 2: I'm not sure I understand. Is the below a different example?

Author: I think he means that the fields prop can also accept an `array`, instead of just a `string`, which is not shown in the example.

So the example can be updated like so:

```vue
<template>
    <h1 :data-directus="setAttr({ collection: 'posts', item: post.id, fields: ['title', 'subtitle', 'slug']
    mode: 'popover' })">
        {{ post.title }}
    </h1>
</template>
```

```html
<h1 data-directus="collection:posts;item:12;fields:title,subtitle,slug;mode:popover">
  I Love Visual Editing
</h1>
```

We can either update the example that is there, or add in a second one showing both ways (`string` or `array`), showing both might take up a lot of space as is not really necessary?

```html
<h1 data-directus="collection:posts;item:12;fields:title;mode:popover">
  I Love Visual Editing
</h1>
```

| Option | Type | Description |
| ------ | ---- | ----------- |
| `collection`  | `string`                           | Name of the relevant collection. This is required. |
| `item`        | `string`, `number`                 | Primary key of the item. This is required. |
| `fields`      | `string`, `string[]`               | The specific fields to show when editing. Optional, otherwise all fields will be shown. |
| `mode`        | `'drawer'`, `'modal'`, `'popover'` | Determines how the edit field(s) should be rendered. Optional, but defaults to `'drawer'` |

::callout{icon="material-symbols:warning-rounded" color="amber"}
**Be careful with sensitive data in attributes**  

> It is recommended that you conditionally render sensitive `data-directus` attributes only when your webpage renders within the Directus Studio Visual Editor in order to avoid leaking this data publicly. <br><br> There are a number of ways to achieve this. For example you might develop your website to only render sensitive attributes in the presence of a certain query string included in your Visual Editor URL settings, for example `?visual-editing=true&token=123`.

Reviewer: Instead of "Directus Studio Visual Editor" only "Directus Visual Editor"?

Reviewer 2: Hmm. I do agree. We don't call it Directus Studio Flows

Reviewer 2: ```suggestion
It is recommended that you conditionally render sensitive `data-directus` attributes only when your webpage renders within the Directus Visual Editor in order to avoid leaking this data publicly. <br><br> There are a number of ways to achieve this. For example you might develop your website to only render sensitive attributes in the presence of a certain query string included in your Visual Editor URL settings, for example `?visual-editing=true&token=123`.
```

::

## Connecting to Directus

> Once all your `data-directus` attributes have been configured, it is required to call the `apply()` method somewhere on the page. This method will establish the connection to Directus and make your editable elements interactive within the Visual Editor Module. This must be done only after all the relevant elements on the page have fully mounted with their generated attributes.

Reviewer: ```suggestion
Once all your `data-directus` attributes have been configured, it is required to call the `apply()` method. This method will establish the connection to Directus and make your editable elements interactive within the Visual Editor Module. This must be done only after all the relevant elements on the page have fully mounted/rendered with their generated attributes.
```

With PHP it is just rendered, not mounted.

Reviewer 2: ```suggestion
Once all your `data-directus` attributes have been configured, it is required to call the `apply()` method somewhere on the page. This method will establish the connection to Directus and make your editable elements interactive within the Visual Editor Module. This must be done only after all the relevant elements on the page have fully mounted or rendered with their generated attributes.
```

Reviewer 2: What do you think of replacing the slash with an or?

```js
apply({ directusUrl: 'http://localhost:8000' });
```

| Option | Type | Description |
| ------ | ---- | ----------- |
| `directusUrl` | `string`                           | URL of your Directus instance. This is required. |

> | `elements`    | `HTMLElement`, `HTMLElement[]`     | Could have one or more elements. If the elements themselves don’t contain a data-directus attribute, their children will be selected. Optional, otherwise will be applied to all elements on page. |

Reviewer: @hola-soy-milk My original writing:

> elements: Optional HTMLElement | HTMLElement[]. Could have one or more elements. If the elements themselves don’t contain a data-directus attribute, their children will be selected.

> Once you specify elements, the options can’t be overwritten anymore.
    You can use elements to scope a section of your page and apply different options than you might have applied to a global selection that already includes those elements. The returned object could then also be used to disable/enable or remove these scoped elements separately from the other elements.

Reviewer 2: I worry about this being too long. [How about footnotes](https://directus.io/docs/configuration/cache)?

Author: Yeah I would avoid putting too much in here as these tables are already really stacked. Footnote can work. Plus, the thing I left out (the ability to scope out sections of the page) one could theoretically infer from understanding.

Reviewer: My intention with this comment was to give you context. It was not meant to use this text as is.

> | `customClass` | `string[]`                         | Adds a class to Overlay Elements to enable custom styles. Optional.|

Reviewer: **Type is string!**

```suggestion
| `customClass` | `string` | Adds a class to Overlay Elements that you can use for custom styling or z-index adjustment. Optional.|
```

Reviewer 2: Is the class called `selected`?

Author: I think he maybe means "specific", ie elements you have selected? @formfcw

Reviewer: I mean "selected". The class is added to the overlay elements of the selected HTML elements, which were selected by the apply-function. In other words, the class is only applied to overlay elements (rectangles) that were selected/found by the apply function. If no `elements` parameter is passed, it means that all overlay elements get that class. If the `elements` parameter is passed, only those overlay elements get that class. Makes sense?

> | `onSaved`     | `(data) => void`                   | Callback function called after the fields are saved in Directus. Optional, otherwise current page will be reloaded using `window.location.reload()` on save. |

Reviewer: Type. Also removed "on save".

```suggestion
| `onSaved` | `(data: { collection: string; item: string | number; payload: Record<string, any> }) => void` | Callback function called after the fields are saved in Directus. Optional, otherwise current page will be reloaded using `window.location.reload()`. |
```

Reviewer 2: Why remove on save?

Author: Yeah im not sure i understand what you mean @formfcw

Reviewer: It is not reloaded "on" save, but after the item is saved. I thought we could just remove it since we already have the context in the sentence before.

> 

Reviewer: Can we somehow show an example of advanced usage like:

```js
const { disable, enable, remove } = apply({ directusUrl: 'http://localhost:8000', elements: titleEl, customClass: 'my-class' , onSaved: () => refresh() });
```

Author: Yeah, I'd add it as an OR underneath the current example so we have both maybe

The `apply()` method also returns the `remove`, `enable` and `disable` methods. These can then be used on the selected elements in question. Make sure to await them.

```js

> const { remove, enable, disable } = await apply({ directusUrl });

Reviewer: in alphabetical order

```suggestion
const { disable, enable, remove } = await apply({ directusUrl });
```

```

> 

Reviewer: Don’t forget to mention that the global `disable()` method (that is imported by the package) also returns an `enable` function to (re-)enable the previously disabled elements.

The optional `onSaved` callback function of the `apply()` method has a number of parameters which may be useful.

| Parameter | Type | Description |
| --------- | ---- | ----------- |
| `collection` | `string`              | Name of the relevant collection. |

> | `collection` | `string`, `number`    | Primary key of the item |

Reviewer: ```suggestion
| `item` | `string`, `number` | Primary key of the item |
```

> | `collection` | `Record<string, any>` | The changed values. |

Reviewer: ```suggestion
| `payload` | `Record<string, any>` | The changed values. |
```

> ## Content Security Policy

Reviewer: This gets too much attention IMO. I suggest putting it in a callout block. It is just a side note.

Reviewer 2: Will extract this into a callout block

> If you have CSP configured, be sure to add `'frame-ancestors': ["'self'", process.env.DIRECTUS_URL]`, to your policy — in other words, make your site available for use inside an iframe in Directus.

Reviewer: `'frame-ancestors': ["'self'", process.env.DIRECTUS_URL]` is Nuxt specific

Reviewer 2: A callout within a callout! 😂

Reviewer: What I tried to say is, that we should not add a code example here. Either we neutralize the CSP rule here or we just say: If you have CSP configured, be sure to make your site available for use inside an iframe in Directus.
